docs/adr/ADR-014-three-layer-architecture-for-external-apis.md

```diff
@@ -0,0 +1,418 @@
+# ADR-014: Three-Layer Architecture for External API Integrations
+
+**Status:** Accepted
+**Date:** 2025-11-14
+**Authors:** cassing
+**Supersedes:** N/A
+**Superseded by:** N/A
+
+## Context
+
+When integrating with external APIs (REST, GraphQL, etc.) in multi-process environments like FastAPI + Celery, we face several challenges:
+
+**Current Situation:**
+
+- Monolithic API client classes mix API communication, caching, and business logic
+- Each process has its own in-memory cache with no sharing across workers
+- No TTL support leads to stale cached data
+- No distributed locking causes race conditions during concurrent cache updates
+- Tight coupling makes testing and mocking difficult
+
+**Problems:**
+
+- **No shared cache:** API workers and background workers can't share cached data
+- **Stale data:** Cached data never expires automatically
+- **Race conditions:** Multiple processes updating cache simultaneously causes corruption
+- **Mixed concerns:** API client handles communication, caching, and business logic
+- **Poor testability:** Hard to mock individual components
+
+**Requirements:**
+
+- Shared cache across multiple processes (FastAPI workers + Celery workers)
+- Distributed locking for thread-safe cache updates
+- TTL support to automatically expire stale data
+- Clear separation between API communication, caching, and business logic
+- Type safety with Pydantic models
+- Easy to test and mock at each layer
+
+**Constraints:**
+
+- Must work in multi-process environment (no shared memory)
+- Must minimize external API calls (rate limits, latency)
+- Must be thread-safe across distributed processes
+- Backward compatible migration path for existing code
+
+## Decision
+
+We adopt a **three-layer architecture** that separates concerns for external API integrations:
+
+**Layer 1: API Client (Pure Communication)**
+
+- Stateless wrapper around external API
+- Handles HTTP/gRPC/GraphQL calls, retries, errors
+- Returns typed Pydantic models
+- No caching, no state, no business logic
+
+**Layer 2: Workspace Client (Cache + Compute)**
+
+- Manages distributed cache with TTL
+- Implements distributed locking for thread-safety
+- Provides compute helpers (filtering, lookups)
+- Uses Layer 1 for API communication
+
+**Layer 3: Service (Business Logic)**
+
+- Implements use cases and orchestration
+- Uses Layer 2 for data access
+- Handles dry-run mode, state transitions
+- Pure business logic, no direct API calls
```

Comment:
To be consistent with usual web framework:

* Layer 1: external client wrapper
* Layer 2: Repository
  * handle all entity operations
* Layer 3: Service
  * handle all domain model operations